moonraker/plugins/history.py

```diff
@@ -0,0 +1,253 @@
+import json, logging, time
+from tornado.ioloop import IOLoop
+
+SAVE_INTERVAL = 5
+
+HIST_NAMESPACE = "history"
+METADATA_KEYS = ['estimated_time','modified','slicer','slicer_version']
+
+class History:
+    def __init__(self, config):
+        self.server = config.get_server()
+        self.database = self.server.lookup_plugin("database")
+        self.gcdb = self.database.wrap_namespace("gcode_metadata",
+            parse_keys=False)
+        self.current_job = None
+        self.file_metadata = None
+        self.jobs = {}
+        self.job_id = -1
+        self.last_update_time = 0
+        self.print_stats = {}
+
+        self.server.register_event_handler(
+            "server:klippy_ready", self._init_ready)
+        self.server.register_event_handler(
+            "server:status_update", self._status_update)
+
+
+        self.server.register_endpoint(
+            "/printer/history/list", ['GET'], self._handle_jobs_list)
+        self.server.register_endpoint(
+            "/printer/history/delete", ['DELETE'], self._handle_job_delete)
+
+        self.database.register_local_namespace(HIST_NAMESPACE)
+        try:
+            self.job_id = self.database.get_item(HIST_NAMESPACE,
+                "job_auto_inc_id")
+        except:
+            logging.info("Creating job history namespace in database")
+            self.database.insert_item(HIST_NAMESPACE,"job_auto_inc_id", 0)
+            self.job_id = 0
+            self.database.insert_item(HIST_NAMESPACE,"prints", self.jobs)
+
+        jobs = self.database.get_item(HIST_NAMESPACE,"prints")
+        for job in jobs:
+            self.jobs[job] = PrinterJob(jobs[job])
+
+    async def _init_ready(self):
+        klippy_apis = self.server.lookup_plugin('klippy_apis')
+        self.print_stats = {}
+
+        try:
+            result = await klippy_apis.query_objects({'print_stats': None})
+        except self.server.error as e:
+            logging.info(f"Error getting print_stats: {e}")
+        self.print_stats = result.get("print_stats", {})
+
+        sub = {"print_stats": None}
+        try:
+            status = await klippy_apis.subscribe_objects(sub)
+        except self.server.error as e:
+            logging.info(f"Error subscribing to print_stats")
+
+    async def _handle_job_delete(self, web_request):
+        args = web_request.get_args()
+        if "all" in args:
+            deljobs = []
+            jobs = list(self.jobs)
+            for job in jobs:
+                self.delete_job(job)
+                deljobs.append(job)
+            return deljobs
+        if "id" not in args:
+            raise self.server.error("No ID to delete")
+        if args['id'] not in self.jobs:
+            raise self.server.error(f"Invalid job id: {args['id']}")
+
+        logging.debug("Jobs %s: %s" % (args['id'], self.jobs))
+        return self.delete_job(args['id'])
+
+    async def _handle_jobs_list(self, web_request):
+        args = web_request.get_args()
```

Comment:
As mentioned above, we should use the web_request's getters to fetch specific arguments rather than get the args dict.  Generally we only want to retrieve the dict if we need to modify it or do some kind of special parsing that isn't available through the standard getters.